File: src/lib/arabic.js
const DIACRITICS_REGEX = /[\u064B-\u065F\u0670\u06D6-\u06ED]/g
const TATWEEL_REGEX = /\u0640/g
const NON_ARABIC_WORD_GAP_REGEX = /[^\p{L}\p{N}\s]+/gu
const MULTISPACE_REGEX = /\s+/g

export function normalizeArabicText(input) {
  return input
    .trim()
    .replace(DIACRITICS_REGEX, '')
    .replace(TATWEEL_REGEX, '')
    .replace(/[إأآٱ]/g, 'ا')
    .replace(/ى/g, 'ي')
    .replace(/ؤ/g, 'و')
    .replace(/ئ/g, 'ي')
    .replace(/ة/g, 'ه')
    .replace(NON_ARABIC_WORD_GAP_REGEX, ' ')
    .replace(MULTISPACE_REGEX, ' ')
    .trim()
}

export function slugifyLabel(input) {
  return input
    .trim()
    .toLowerCase()
    .replace(/[^\p{L}\p{N}\s-]+/gu, '')
    .replace(MULTISPACE_REGEX, '-')
}

File: backend/arabic.py
import re


DIACRITICS_REGEX = re.compile(r"[\u064B-\u065F\u0670\u06D6-\u06ED]")
NON_WORD_GAP_REGEX = re.compile(r"[^\w\s]+", re.UNICODE)
MULTISPACE_REGEX = re.compile(r"\s+")


def normalize_arabic_text(value: str) -> str:
    normalized = value.strip()
    normalized = DIACRITICS_REGEX.sub("", normalized)
    normalized = normalized.replace("\u0640", "")
    normalized = re.sub(r"[إأآٱ]", "ا", normalized)
    normalized = normalized.replace("ى", "ي")
    normalized = normalized.replace("ؤ", "و")
    normalized = normalized.replace("ئ", "ي")
    normalized = normalized.replace("ة", "ه")
    normalized = NON_WORD_GAP_REGEX.sub(" ", normalized)
    normalized = MULTISPACE_REGEX.sub(" ", normalized)
    return normalized.strip()


def slugify_label(value: str) -> str:
    lowered = value.strip().lower()
    cleaned = re.sub(r"[^\w\s\u0600-\u06FF-]+", "", lowered, flags=re.UNICODE)
    return MULTISPACE_REGEX.sub("-", cleaned).strip("-")
